SecurityPkg/Tcg/Tcg2Dxe/Tcg2Dxe.c

```diff
@@ -2797,5 +2978,11 @@ DriverEntry (
   Status = InstallTcg2 ();
   DEBUG ((EFI_D_INFO, "InstallTcg2 - %r\n", Status));
 
+  //
+  // Set TPM2 ACPI table
+  //
+  Status = PublishTpm2 ();

```

Comment:
EDK is inconsistent :)  Unlike Tcg2Dxe which left installation of ACPI table to Tcg2Smm, TcgDxe publishes ACPI table on its own.  Weirdly enough TcgSmm also publishes some ACPI table which it reads from FV, not sure why.

Reply:
@SergiiDmytruk any update here?

Reply:
What for?  There are no changes for TPM1.2 because they aren't needed: TcgDxe already publishes the log.

Just checked it again to be sure and it does, although when coreboot supports both TPM versions TPM2.0 log format is used and that makes EDK to fail importing the log.  If we ship boards that support both TPM versions either need to pick log version dynamically in  coreboot or import truncated SHA-256 hashes from TPM2 log in EDK (because there are no SHA-1 hashes).  Otherwise event log will miss coreboot part when TPM1 is in use.

Reply:
What I mean is that we skip publishing TCPA table if we have TPM1.2: https://github.com/Dasharo/coreboot/pull/517/files#diff-28cac0ee6b199882d2ecb787e057e226f4a84e73bd5e4900d5fd1b5421afef26R223

IF you make modifications only to TCG2 modules, we will end up without TCPA table (event logs aside)

Reply:
> And it doesn't need any modifications?

It's just SHA-1 all the time for TPM1, so log entries from coreboot and EDK already align with each other unless you want to drop trailing zeros from events coming from coreboot (not done for TPM2 either).

> I would expect coreboot to create an appropriate log depending on detected TPM

You brought that up during review [here](https://review.coreboot.org/c/coreboot/+/68747/comment/cbfd1fd6_db1c9692/).  I guess there could be a separate `TPM_LOG_TCG` that picks log format automatically.  On Talos TPM and log versions didn't align with each other, I think that's why log format needed to be independent.

Reply:
OK, I get it, that sometimes you may want to force a log format.

So what will happen when coreboot creates the TPM2-compliant log format for TPM1.2? Will EDK2 actually use the same format in TcgDxe?

Reply:
> So what will happen when coreboot creates the TPM2-compliant log format for TPM1.2?

coreboot's event log won't be imported.  Also because TcgDxe doesn't create its own Event ID entry, there will be none.

Reply:
Sorry, but I still don't understand your previous statement:
`What for? There are no changes for TPM1.2 because they aren't needed: TcgDxe already publishes the log.`
Together with this: `coreboot's event log won't be imported. Also because TcgDxe doesn't create its own Event ID entry, there will be none.` I'm unable to be convinced that there are no modifications required

Reply:
No modification required for making EDK's TPM1.2 event log available to an OS.  Changing what's published requires modifications here or in coreboot if we want to support those scenarios (picking TPM1 or TPM2 dynamically).

Reply:
>  if we want to support those scenarios (picking TPM1 or TPM2 dynamically).

That was the goal of the whole effort of unifying TPM drivers and eventlogs. To get the right driver talk to TPM and have coreboot create an appropriate log for the detected TPM, rather than rely on fixed CONFIG.

While this PR addresses the problem of the missing log entries from coreboot, I would also want to see it working correctly in the dynamic scenarios.
@macpijan cc

I can close this thread if I see some next steps taken in that direction (i.e., tasks/PRs).

Reply:
Looks like I almost have the changes for enabling dynamic log selection in coreboot.  It wasn't hard to do them, but original code for registering probing functions (not the one which landed in upstream) didn't work in some stages and the log was missing stuff.  Now things work for TPM1 fine but there is somehow no log in OS for TPM2.  Can send a coreboot PR after sorting it out.

Reply:
> Now things work for TPM1 fine but there is somehow no log in OS for TPM2.

Seems like an issue on Q35 or with Qemu/swtpm because the same happens without new changes and everything works fine on APU.

I've added the changes to https://github.com/Dasharo/coreboot/pull/517.  There is a side-effect when new option is on: probing for TPM in basically every stage which needs to know algorithm for hashing data because the algorithm depends on TPM now.